src/utils.ts

```diff
@@ -222,3 +223,64 @@ export async function retry<T>(call: () => Promise<T>, times: number, delayS: nu
     });
   return promiseChain;
 }
+
+/**
+ * Estimates the total gas cost required to submit an unsigned (populated) transaction on-chain
+ * @param unsignedTx The unsigned transaction that this function will estimate
+ * @param relayerAddress The address that the transaction will be submitted from ()
+ * @param provider A valid ethers provider - will be used to reason the gas price
+ * @param gasPrice A manually provided gas price - if set, this function will not resolve the current gas price
+ * @returns The total gas cost to submit this transaction - i.e. gasPrice * estimatedGasUnits
+ */
+export async function estimateTotalGasRequiredByUnsignedTransaction(
+  unsignedTx: PopulatedTransaction,
+  relayerAddress: string,
+  provider: providers.Provider | L2Provider<providers.Provider>,
+  gasPrice?: BigNumberish
+): Promise<BigNumberish> {
+  const voidSigner = new VoidSigner(relayerAddress, provider);
+  // Verify if this provider has been L2Provider wrapped
+  if (isL2Provider(provider)) {
+    const populatedTransaction = await voidSigner.populateTransaction(unsignedTx);
+    return (await provider.estimateTotalGasCost(populatedTransaction)).toString();
+  } else {
+    // Estimate the Gas units required to submit this transaction
+    const estimatedGasUnits = await voidSigner.estimateGas(unsignedTx);
+    // Provide a default gas price of the market rate if this condition has not been set
+    const resolvedGasPrice = gasPrice ?? (await provider.getGasPrice());
+    // Find the total gas cost by taking the product of the gas
+    // price & the estimated number of gas units needed
+    return BigNumber.from(resolvedGasPrice).mul(estimatedGasUnits).toString();
+  }
+}
+
+/**
+ * Create an unsigned transaction of a fillRelay contract call
+ * @param spokePool The specific spokepool that will populate this tx
+ * @param destinationTokenAddress A valid ERC20 token (system-wide default is UDSC)
+ * @param simulatedRelayerAddress The relayer address that relays this transaction
+ * @returns A populated (but unsigned) transaction that can be signed/sent or used for estimating gas costs
+ */
+export async function createUnsignedFillRelayTransaction(
+  spokePool: BaseContract,
```

Comment:
nit: you didn't originally design it this way, but I see that each implementation uses PolygonSpokePool, OptimismSpokePool, etc. However, the interfaces we're using across these classes come from the base SpokePool class.

Anyway, to get better typing, we could change all of those to just SpokePool and take that type in here. Might make things a little cleaner and give us better type-checking.

Reply:
100% agree. I've updated the files to use SpokePool instead of BaseContract [here](https://github.com/across-protocol/sdk-v2/pull/64/commits/ede380a86add898ec16ee34c058fd156eae7a086).